V5 add back support for node builtins (#11764)
facebook/create-react-app: docusaurus/docs/nodejs-builtin-fallbacks.md

---
id: nodejs-builtin-fallbacks
title: NodeJS builtin fallbacks
---

NodeJS builtin fallbacks enable you to import NodeJS builtin modules meant for Node and fallback to browser specific modules in your web application.

> Per default Create React App set fallbacks to empty modules in production build and development fallbacks in development mode.

Reviewer: ```suggestion
By default, Create React App set fallbacks to empty modules in the production build and development fallbacks when in development mode.
```

> Per default Create React App set fallbacks to empty modules in production build and development fallbacks in development mode.

Reviewer 2: ```suggestion
This restores the previous behavior in Webpack < 5, where Create React App used to include these builtin fallbacks.
```

I think this is how the sentence was intended, connecting to the previous sentence. I changed the inclusion to past tense.

> In development mode you might get error messages in your browser console similar to:

Reviewer: ```suggestion
In development mode, you receive error messages in the terminal where you're running the development server similar to:
```

i'm seeing the errors in my terminal where i'm running `npm start`, not browser console

```

> (dev) Error: Module "path" not found, cannot access property "join", please read https://create-react-app.dev/docs/nodejs-builtin-fallbacks

Reviewer 3: this appears to be a broken link

```

> _(It's possible to disable these warnings in development using the environment variable: `DISABLE_MISSING_NODEJS_BUILTIN_MODULE_FALLBACK_WARNING=true`)_

Reviewer: ```suggestion
_You can disable these warnings in development using the environment variable: `DISABLE_MISSING_NODEJS_BUILTIN_MODULE_FALLBACK_WARNING=true`_
```

**IMPORTANT:** Before fixing this dependency, please make sure to only use Npm packages meant for the browser and not for Node / backend.

> It takes abit of work but visit the project documentation, README.md and if on GitHub etc. look open and closed issues in the project e.g. search "browser" to see if maintainers close issues for browser support.

Reviewer: ```suggestion
It takes a bit of work but visit the project documentation, README.md and if on GitHub etc. look open and closed issues in the project e.g. search "browser" to see if maintainers close issues for browser support.
```

> Implications of loading packages not build for the browser can vary from security, bundle size etc. - There might be better alternatives.

Reviewer: ```suggestion
There can be unexpected implications of loading packages not build for the browser which include security, bundle size etc. It is generally a better to use packages intended for the browser if possible.
```

> **Escape hatch**

Reviewer: ```suggestion
## Escape hatch
```

> To fix the issue you will need to add the browser fallback - the example above complains about missing `path` module.

Reviewer: ```suggestion
If you cannot find a native browser replacement module, you can use one of the following polyfills. 
You just have to install the module, and Create React App will use it. 

For example, if you're seeing the above mentioned example for the `path` module: 
```

```bash
npm install path-browserif
# or..
yarn add path-browserif
```

Create React App will recognize the fallback and use that instead of an empty module.
_(Find the fallback package in the table bellow)_

| NodeJS builtin module | Browser version            |
| :-------------------- | :------------------------- |
| assert                | assert                     |
| buffer                | buffer                     |
| console               | console-browserif          |
| constants             | constants-browserif        |
| crypto                | crypto-browserif           |
| domain                | domain-browse              |
| events                | events                     |
| http                  | stream-htt                 |
| https                 | https-browserif            |
| os                    | os-browserify/browse       |
| path                  | path-browserif             |
| punycode              | punycode                   |

> | process               | process/browse             |

Reviewer 4: Typo: should be `process/browser` (missing r at the end).

Reviewer 5: At least half the names in the 2nd column seem to be missing their last letter:
* `browserif` should be `browserify` ? (multiple variants)
* `stream-htt` should be `stream-http` ?
* `readable-stream/duple` should be `readable-stream/duplex` ?
etc etc etc

Reviewer 6: We're looking at this at the moment, and just noticed this issue too. Thanks @no23reason and @PjotrB.
